File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::unordered_set<char> char_set;
        int l = 0;
        int r = 0;
        int max_len = 0;
        int n = s.size();
        while(r < n ){
            while(char_set.find(s[r])!=char_set.end()){
                char_set.erase(s[l]);
                l++;
            }
            if (char_set.find(s[r])==char_set.end()){
                char_set.insert(s[r]);
                r++;
            }
            max_len = max((int)char_set.size(),max_len);
            
        }
        return max_len;
        
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charSet = set()
        l = 0
        r = 0
        max_len = 0
        n = len(s)
        while r < n:
            while s[r] in charSet:
                charSet.remove(s[l])
                l+=1
            charSet.add(s[r])
            r+=1
            max_len = max(len(charSet),max_len)
        
        return max_len

        